Lib/glyphsLib/builder/layers.py

```diff
@@ -12,53 +12,83 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from .constants import GLYPHS_PREFIX
+import re
+
+from .constants import (
+    FOREGROUND_LAYER_ID_KEY,
+    GLYPHS_PREFIX,
+    LAYER_ID_KEY,
+    LAYER_ORIGINAL_NAME_KEY,
+)
 
-LAYER_ID_KEY = GLYPHS_PREFIX + "layerId"
 LAYER_ORDER_PREFIX = GLYPHS_PREFIX + "layerOrderInGlyph."
 LAYER_ORDER_TEMP_USER_DATA_KEY = "__layerOrder"
 
 
 def to_ufo_layer(self, glyph, layer):
+    glyphs_layers = self.font.glyphs[glyph.name].layers
     ufo_font = self._sources[layer.associatedMasterId or layer.layerId].font
+
     if layer.associatedMasterId == layer.layerId:
         ufo_layer = ufo_font.layers.defaultLayer
-    elif layer.name not in ufo_font.layers:
-        ufo_layer = ufo_font.newLayer(layer.name)
-    elif layer.name in ufo_font.layers and glyph.name in ufo_font.layers[layer.name]:
-        self.logger.warning(
-            "%s %s: Glyph %s, layer %s: Duplicate glyph layer name",
-            ufo_font.info.familyName,
-            ufo_font.info.styleName,
-            glyph.name,
-            layer.name,
-        )
-        n = 1
-        new_layer_name = layer.name
-        while new_layer_name in ufo_font.layers:
-            new_layer_name = layer.name + " #" + repr(n)
-            n += 1
-        ufo_layer = ufo_font.newLayer(new_layer_name)
     else:
-        ufo_layer = ufo_font.layers[layer.name]
-    if self.minimize_glyphs_diffs:
+        # Find foreground layer and use the same name as a base for the
+        # background layer name.
+        layer_name = next(
+            (
+                l.name
+                for l in ufo_font.layers
+                if l.lib.get(LAYER_ID_KEY) == layer.layerId
```

Comment:
maybe you could store a map from layerId to layer name in the builder instance instead of finding them like this each time